### ISSUE TITLE
Wrong warning: `value assigned never read` because used in return

### ISSUE THREAD

**Issue by tforgione** [rust-lang/rust#50075]
I admit this code looks stupid, but it's my minimal working example:

``` rust
fn compute_test() -> u32 {
    let mut test: u32 = 0;
    test = 1;
    test
}

fn main() {
    println!("{}", compute_test());
}
```

[This snippet](https://play.rust-lang.org/?gist=c2f735f23e616a0846aca07d5a263bd1&version=nightly) compiles and give the following warning:
```
warning: value assigned to `test` is never read
 --> src/main.rs:2:9
  |
2 |     let mut test: u32 = 0;
  |         ^^^^^^^^
  |
```

even though it's being used in the return statement. I feel like this warning is kind of missleading.

**Comment by jonas-schievink** [comment]
The warning explicitly mentions the *value*, which is indeed not used because you overwrite it in the next line.

**Comment by jonas-schievink** [comment]
Perhaps the span should be moved to actually highlight the value and not the binding, then.
